wallet-new/docs/updates.md

```diff
@@ -0,0 +1,563 @@
+            type: integer
+        type: object
+
+# Specification<a id="sec-5" name="sec-5"></a>
+
+## New update system<a id="sec-5-1" name="sec-5-1"></a>
+
+**Summary**
+
+`We propose changes to the wallet backend data model and the way the installers interact with the cardano launcher.`
+
+To begin with, we will stop downloading an update in the background, but rather we will give the user full control.
+When requested, an update will be downloaded in a transient directory and when the user is ready to apply it,
+the downloaded installer will be moved to the folder the Cardano launcher currently watches, so that the
+installation process can be reused without changes to the underlying infrastructure.
+
+We propose adding some extra fields to what used to be called `CUpdateInfo` (which here we rename for
+clarity):
+
+    data WalletSoftwareUpdate = WalletSoftwareUpdate
+        { wsuSoftwareVersion :: !SoftwareVersion
+        , wsuBlockVesion     :: !BlockVersion
+        , wsuScriptVersion   :: !ScriptVersion
+        , wsuImplicit        :: !Bool
+        , wsuVotesFor        :: !Int
+        , wsuVotesAgainst    :: !Int
+        , wsuPositiveStake   :: !CCoin
+        , wsuNegativeStake   :: !CCoin
+        , wsuSeverity        :: !UpdateSeverity
+        -- ^ The severity of the update (i.e. optional or mandatory).
+        , wsuUpdateState     :: !UpdateState
+        -- ^ The state of this update from users point of view.
+        , wsuSha256Checksum  :: !Sha256Checksum
+        -- ^ The checksum expected for this installer. It will probably
+        -- need to be passed as part of the `UpdateProposal` (maybe in an attribute?),
+        -- or we could reuse the hash which is part of the name, if possible.
+        } deriving (Eq, Show, Generic, Typeable)
+
+where `UpdateSeverity` is defined as an enumeration indicating how compulsory this update is:
+
+    data UpdateSeverity = Mandatory
+                        | Optional
+
+For hard forks updates we would set the severity to `Mandatory`, for soft forks and other non-critical updates
+we would label an update as `Optional`.
+
+-   `Optional` updates can be rejected by the user, as well as applied to will at a later stage;
+-   `Mandatory` updates will need to be applied immediately and won't allow the user to proceed
+    in using Daedalus.
+
+On the other hand, `wsuUpdateState` records the users' decision towards the update:
+
+    data UpdateState = Applied
+                     -- ^ The update has been fully applied.
+                     | Downloaded !Sha256Checksum !FilePath
+                     -- ^ The update has been downloaded but not applied yet. A Checksum
+                     -- can be stored to verify the integrity of the download.
+                     | Skipped
+                     -- ^ The update has been skipped by the user.
+                     | Proposed
+                     -- ^ The update has been just proposed and is waiting for users' feedback.
```

Comment:
Shouldn't we add `Downloading` state here? An update which is currently in process of downloading doesn't seem to fit in any of those categories.

Reply:
I thought about this, but consider the case where the user is downloading the update but it hard-quits Daedalus. If we store this "transient" state, it will create troubles when we re-access the app, because it would look like the update is downloading, but we really want to start fresh and pretend we never downloaded the upgrade, so I would be tempted to keep `Downloading` out of the sum type and review this in case that's not the right choice.